File: leetcode/golang/LCR_154_fu_za_lian_biao_de_fu_zhi_lcof.go
type Node struct {
    Val int
    Next *Node
    Random *Node
}

   
func copyRandomList(head *Node) *Node {
    tmp := head
    var hashMap = make(map[*Node]*Node)
    for tmp != nil {
        hashMap[tmp] = &Node{
            Val: tmp.Val, 
        }
        tmp = tmp.Next
    }
    tmp = head
    for tmp != nil {
        hashMap[tmp].Next = hashMap[tmp.Next]
        hashMap[tmp].Random = hashMap[tmp.Random]
        tmp = tmp.Next
    }
    return hashMap[head]
}

File: leetcode/c++/LCR_154_fu_za_lian_biao_de_fu_zhi_lcof.cpp
// https://leetcode.cn/problems/fu-za-lian-biao-de-fu-zhi-lcof/description

#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*, Node*> map;
        Node* tmp = head;
        while (tmp) {
            map[tmp] = new Node(tmp->val);
            tmp = tmp->next;
        }
        tmp = head;
        while (tmp) {
            map[tmp]->next = map[tmp->next];
            map[tmp]->random = map[tmp->random];
            tmp = tmp->next;
        }
        return map[head];
    }
};

// 时间复杂度O(n)，空间复杂度O(n)